tests/flow/test_bidirectional_traversals.py

```diff
@@ -0,0 +1,258 @@
+    def test03_bidirectional_variable_length_traversals(self):
+        query = """MATCH (a)-[*]-(b) RETURN a.val, b.val ORDER BY a.val, b.val"""
+        actual_result = redis_graph.query(query)
+        # Each combination of distinct node source and destination should appear once.
+        expected_result = [['v1', 'v2'],
+                           ['v1', 'v3'],
+                           ['v2', 'v1'],
+                           ['v2', 'v3'],
+                           ['v3', 'v1'],
+                           ['v3', 'v2']]
+        self.env.assertEquals(actual_result.result_set, expected_result)
+
+        # Should generate the same results as the previous query.
+        query = """MATCH (a)-[*1..2]-(b) RETURN a.val, b.val ORDER BY a.val, b.val"""
+        actual_result = redis_graph.query(query)
+        self.env.assertEquals(actual_result.result_set, expected_result)
+
+    # Test collecting self and all direct neighbors.
+    def test04_bidirectional_variable_bounded_length_traversals(self):
+        query = """MATCH (a)-[*0..1]-(b) RETURN a.val, b.val ORDER BY a.val, b.val"""
+        actual_result = redis_graph.query(query)
+        # Each combination of distinct node source and destination should appear once.
+        expected_result = [['v1', 'v1'],
+                           ['v1', 'v2'],
+                           ['v2', 'v1'],
+                           ['v2', 'v2'],
+                           ['v2', 'v3'],
+                           ['v3', 'v2'],
+                           ['v3', 'v3']]
+        self.env.assertEquals(actual_result.result_set, expected_result)
+
+    # Test bidirectional query on nonexistent edge.
+    def test05_bidirectional_variable_length_traversals_over_nonexistent_type(self):
+        query = """MATCH (a)-[:NONEXISTENT*]-(b) RETURN a.val, b.val ORDER BY a.val, b.val"""
+        actual_result = redis_graph.query(query)
+        expected_result = []
+        self.env.assertEquals(actual_result.result_set, expected_result)
+
+    # Test bidirectional query on real edge or nonexistent edge.
+    def test06_bidirectional_variable_length_traversals_over_partial_existing_types(self):
+        query = """MATCH (a)-[:NONEXISTENT|:E*]-(b) RETURN a.val, b.val ORDER BY a.val, b.val"""
+        actual_result = redis_graph.query(query)
+        # Each combination of distinct node source and destination should appear once.
+        expected_result = [['v1', 'v2'],
+                           ['v1', 'v3'],
+                           ['v2', 'v1'],
+                           ['v2', 'v3'],
+                           ['v3', 'v1'],
+                           ['v3', 'v2']]
+        self.env.assertEquals(actual_result.result_set, expected_result)
+
+    # TODO returns 16 rows; 18 rows expected.
+    # The missing two rows are both `['v2', 'v3']
+    # Test bidirectional query on two real edge types.
+    #  def test07_bidirectional_variable_length_traversals_over_multiple_existing_types(self):
```

Comment:
I'd like to enable these tests once we switch to edge uniqueness instead of node uniqueness, but if you prefer I can delete these lines and add them once that's done? Your call!